### ISSUE TITLE
add ecosystem integration to jiff? (e.g., optional dependencies on `icu`, `diesel`, `sqlx`, etc)

### ISSUE THREAD

**Issue by BurntSushi** [BurntSushi/jiff#50]
In #4 and in other places, I've resisted adding "ecosystem integration" into Jiff. Not necessarily because I expected others to instead depend on Jiff necessarily right away, but that _eventually_, as folks started using Jiff (hopefully), it would make sense for them to add a Jiff option.

But I think I should probably flip this script. Getting Jiff adopted is going to be difficult, and I don't know if it has any broadly applicable "killer feature" that will convince folks to jump over from other datetime libraries. (I think Jiff's `Span` is really amazing, but I haven't figured out the best way of communicating that to a broad audience yet.) I think this means that the best path to Jiff adoption is to reduce friction as much as possible.

I get the sense that folks rely heavily on other crates providing integration points for their chosen datetime types of choice. In particular, it is often the case that converting from one type to another is actually quite tricky. Even tricky enough that [I myself get tripped up](https://github.com/sfackler/rust-postgres/pull/1164). So expecting users to just hand-roll these sorts of integrations is probably pretty lofty.

So... I think one way to reduce friction is for Jiff to provide these integrations. One of my initial reservations was introducing public dependencies on crates that are "less stable/mature." While Jiff is currently brand new and I expect to do some breaking change releases, I do hope that in about ~1 year's time, we'll have a Jiff 1.0 and that I'll commit to that for the foreseeable future with no breaking changes. However, I actually don't think this is a huge deal. We can just include the major/minor version of the crate in the feature name itself, [just like `rust-postgres` does](https://github.com/sfackler/rust-postgres/blob/d2634a46b5bc1de5176f0766626e7a9f6757f362/postgres-types/Cargo.toml#L53-L54).

I have two concerns with this approach. That is, if I start this way by introducing dependencies on crates like `diesel`, then:

1. Does this mean it will become very difficult, if not impossible, to coordinate a switch where instead of Jiff providing the integration points, the higher level libraries do it instead?
2. A datetime library ought to appear "low" in the dependency tree. It is intended to be a building block for other systems. So if Jiff starts depending on "bigger" systems like `diesel` (`diesel` is just an example, this applies to most crates I think), then what happens if those bigger systems actually need to depend on a datetime library? Recursive dependencies are a no-go.

Should I pursue this course? Should I limit it to Jiff 0.x versions and then, once 1.0 is out, insist that the integration points get flipped? I'm not sure. That will make 3 datetime choices for a lot of crates, and it's a tough sell.

**Comment by lu-zero** [comment]
Can you make those integrations a separate crate?

**Comment by BurntSushi** [comment]
Not easily. Or at least, not without a degraded user experience. Because of the orphan rule. That is, I expect most "integrations" to be implementing traits for Jiff data types. That means the integration needs to be:

1. In the place where the trait is defined. (So, e.g., in diesel.)
2. In the place where the type is defined. That is, in `jiff`.
3. In some other crate with a newtype wrapper. Then anyone using this crate has to manage the newtype wrapper. This is "doable," but there's friction involved. It can be lessened using various tricks, but I don't think it can be zero friction.

**Comment by sslivkoff** [comment]
for (3) maybe a nice macro could make minimize the friction

this will make dependency-conscious ppl more likely to integrate jiff

**Comment by joshtriplett** [comment]
I don't think `jiff` should have even *optional* dependencies on crates like `sqlx`; having a low-level data type even *optionally* depend on a database in order to implement the database's trait seems a little too absurd, as well as (as you noted) making it difficult for those libraries to (directly or indirectly) depend on `jiff`.

One of these days, I hope we figure out a way to relax the orphan rule enough that a `sqlx-jiff` crate can implement integration between `sqlx` traits and `jiff` types. But we don't have that yet, and still need a solution.

Ideally, some of the crates will be willing to accept PRs early on, in the hopes that they'll treat a new low-level data type crate favorably. If they don't want to integrate it because it's pre-1.0, you might be able to get confirmation that they'd be willing to integrate it once it's 1.0.

If that doesn't work, then for something like `sqlx`, for now, I think it'd be *entirely* reasonable to have a `sqlx-jiff` crate that provides 1) a newtype wrapper, 2) bidirectional `From` impls, 3) a `to_jiff()` method on the newtype, and 4) an extension trait with a `to_sqlx()` method on jiff types, to avoid having to turbofish the `From` impl for disambiguation (because `.into()` is ambiguous when dealing with an "accept anything implementing a trait" crate like `sqlx`). That's not zero-friction, but it's low-friction.

**Comment by BurntSushi** [comment]
I like that plan @joshtriplett. I think that's probably best way to go. I think this will be my next focus, because I feel like I hear "jiff doesn't have as much ecosystem integration" as a big friction point. The newtype approach when an integration point downstream is perhaps not yet desirable does seem better than doing it directly from Jiff. Definitely not ideal, but I think there is enough "meat and potatoes" when converting between datetime types that it is actually worth doing.

**Comment by thumbert** [comment]
Hi, 

If you finished the postgres support for jiff, can you please do the same for DuckDB?  It has exactly the same data types as postgres, so I hope it should be trivial.  The relevant crate is `duckdb-rs`.

Thank you very much!  I am growing my Rust usage and slowly shedding the chrono dependency.

Tony

**Comment by BurntSushi** [comment]
I haven't done anything yet. I'll likely prioritize the crates that folks use the most first. But I'll keep `duckdb-rs` in mind!

**Comment by tisonkun** [comment]
> I think it'd be entirely reasonable to have a sqlx-jiff crate that provides

Here is a snippet I used in my private project that proves this way works - https://github.com/launchbadge/sqlx/issues/3487#issuecomment-2366739587 (wrap Timestamp to bridge from/to PG's TIMESTAMPTZ)

I'm not quite familiar with the wrapper way. If one can provide a code snippet of how we can implement "2) bidirectional From impls, 3) a Deref impl", I may be able to spend some time trying out a sqlx-jiff crate.

It will define feature flags like "postgres", "mysql" and "sqlite" for different backends.

**Comment by BurntSushi** [comment]
I think I would prefer having the integration crates live in this repository. Some of the integrations (like PostgreSQL, as you've discovered) as pretty subtle, and I think it's important to have them under the umbrella of either the Jiff project or the project that's being integrated with.

I don't have a ton of time unfortunately to do abstract mentoring. With that said, if you want to get started by submitting a PR to this repo with a new `jiff-sqlx` crate, I can probably do a review.

**Comment by azerupi** [comment]
Could implementing `From` and `Into` the chrono types give a first simple solution to this problem? 

Obviously it would not allow the user to drop chrono as a dependency, but it could allow the user to work with jiff's API until they have to hand it off to a library that doesn't support jiff yet. The ecosystem already has support for chrono, so it could be a general solution until the ecosystem provides support for jiff directly?

**Comment by BurntSushi** [comment]
I don't think it makes sense in really any circumstance to add a dependency on `chrono` from `jiff`. Even an optional one. I think the route to go there is something like a `jiff-chrono` (or `chrono-jiff`) crate that facilitates conversions to-and-from types between the crates.

It's also not especially straight-forward to just provide `From` impls. It works in some cases. For example, `chrono::DateTime<Utc>` and `Timestamp` make sense. But what about a `Zoned`? You might need `chrono-tz` for that. Honestly, it's all very thorny and I haven't thought through all the intricacies of conversions yet. And if you aren't careful, the conversion step could easily lose information along the way.

It's also not totally clear to me that `From` impls necessarily make anything automatic. I think you still need to insert some kind of explicit conversion code when working with crates that, say, only support Chrono and not Jiff. The `From` impls provide value by having an expert (probably me) implement them, but those can be in a `jiff-chrono` crate (or something).

Ecosystem integration is definitely on my list of things I want to do soon. I can definitely see that it is a blocker for some folks. Or at least, is causing folks to bounce off of Jiff. ("I tried using Jiff, but sqlx didn't support it and I didn't want to dig into it so I just dropped it for now." is something I heard recently.) I'm considering doing a `jiff 0.2` release ~soon, and then that might be a good time to publish some ecosystem integration crates to bootstrap things.

My hope is that once I get Jiff 1.0 out (I'm still planning on doing that in summer of this year, 2025), folks will be more willing to add integration points with it upstream. But we'll see.

**Comment by azerupi** [comment]
> It's also not totally clear to me that From impls necessarily make anything automatic. I think you still need to insert some kind of explicit conversion code when working with crates that, say, only support Chrono and not Jiff.

Yes, the explicit conversion at the boundaries with ecosystem crates was how I imagined it working. To me at least it wouldn't sound too bad to require an explicit conversion for the time being whenever you convert to/from e.g. diesel models.

**Comment by BurntSushi** [comment]
Ah okay, yeah, I think we're on the same page then. I think that falls into a `jiff-chrono` crate (and probably also `jiff-time`). I think those crates are good ideas and I do plan to make them.

**Comment by musicinmybrain** [comment]
> Does this mean it will become very difficult, if not impossible, to coordinate a switch where instead of Jiff providing the integration points, the higher level libraries do it instead?

I think that the example of `rkyv` suggests that if you offer integrations as features in the `jiff` crate itself, it might indeed be quite difficult to smoothly and gracefully stop offering them in the future.

https://github.com/rkyv/rkyv/blob/49262dc189d8d7236c6dfba0c1a19896c2a6e771/rkyv/Cargo.toml#L26-L45

They write,

> Support for various common crates. These are primarily to get users off the ground and build some momentum.
>
> These are NOT PLANNED to remain in rkyv for the 1.0 release. Much like serde, these implementations should be moved into their respective crates over time. Before adding support for another crate, please consider getting rkyv support in the crate instead.

Now, I haven’t been watching *that* closely, but I haven’t noticed any of the “ecosystem” crates actually adding support for `rkyv`. It’s not obvious that any progress will be made on migrating these integrations out of `rkyv` before a 1.0 release eventually comes out and drops them all.

**Comment by BurntSushi** [comment]
Yeah I suspect you're right. `chrono` does notably have an `rkyv` integration, but I agree it is in general uncommon.

**Comment by jaskij** [comment]
Important up front: for better or worse, I'd start with `chrono` - large parts of the ecosystem adopted it as the default time crate. Once that's done, interested people could write PRs to other crates to add support for `jiff` without your input.

----

I'm very sympathetic to what you're trying to achieve, and would *love* to use `jiff`. But with no way to do conversion to and from anything that I could store in the database (Postgres in my case, using raw `tokio-postrgres`), it's just not viable to use.

----

Another point is fallibility: I know that it's not always possible to convert between two types even though both are supposed to represent the same thing. How would that be handled?

As an example: [Protobuf timestamps](https://docs.rs/prost-types/latest/prost_types/struct.Timestamp.html) are not always convertible. Heck, even converting `jiff::Timestamp` to `prost_types::Timestamp` would be fallible, since Protobuf constrains the dates to ones that can be represented as RFC 3339 strings (only allowing years in the range 0-9999 inclusive).

On that note: I could probably write the `prost_types` conversion myself, since the two timestamp types are so close, except Google doesn't use leap seconds (using smear seconds instead) and I don't have a clue how that plays into everything.

----

I would also not put too much stock in releasing version 1.0 being a big blocker for people. For better or worse, Cargo and the Rust ecosystem provide stronger guarantees than the semantic versioning spec, and there's a lot of quite big and popular crates that are still pre-1.0. `rand`, `num`, `sqlx`, `tracing`, to name a few.

**Comment by tisonkun** [comment]
@jaskij 

> Postgres in my case, using raw tokio-postrgres

FYI, tokio-postgres supports jiff types from 0.7.12: https://github.com/sfackler/rust-postgres/commit/6de0fceebe3c1800a5e812f35449e464fbc33f55

**Comment by BurntSushi** [comment]
Doesn't `tokio-postgres` have Jiff support? https://github.com/sfackler/rust-postgres/blob/07b6878d5c72e20ff47be6b618cc2b88c43ff661/tokio-postgres/Cargo.toml#L37

> As an example: [Protobuf timestamps](https://docs.rs/prost-types/latest/prost_types/struct.Timestamp.html) are not always convertible. Heck, even converting jiff::Timestamp to prost_types::Timestamp would be fallible, since Protobuf constrains the dates to ones that can be represented as RFC 3339 strings (only allowing years in the range 0-9999 inclusive).

I think you just treat it as an error like any other right? You're going to have the same problem here with pretty much any other datetime library right?

> On that note: I could probably write the prost_types conversion myself, since the two timestamp types are so close, except Google doesn't use leap seconds (using smear seconds instead) and I don't have a clue how that plays into everything.

Neither does Jiff. So you're good.

> I would also not put too much stock in releasing version 1.0 being a big blocker for people. For better or worse, Cargo and the Rust ecosystem provide stronger guarantees than the semantic versioning spec, and there's a lot of quite big and popular crates that are still pre-1.0. rand, num, sqlx, tracing, to name a few.

It's not about 1.0 per se. It's about getting to a version and committing to it. I want to send a "this is a stable foundation to build upon" message.

We don't really need to debate the meaning of 1.0 here. I've done my fair share of 1.0 releases in my ten years building crates for Rust. (`regex`, `bstr`, `aho-corasick`, `csv`, `walkdir` and probably more that I'm forgetting.)

**Comment by jaskij** [comment]
Thanks for the info about `jiff` support in `postgres-types`, that's a major blocker crossed out for me. Turns out their documentation got updated for `FromSql` but not `ToSql`, and I was looking at the latter.

----

When it comes to fallibility - I just wanted to point it out, since I'd be surprised if there wasn't some naive crate somewhere that treats this conversions as infallible. But as long as it's not widespread, that's not an issue for you.

----

Re: 1.0, again, just wanted to make a point, but you have it all well thought it and it's not my intention to force you to change your view.
